File: Crash it/MovingItem.swift
//
//  MovingItem.swift
//  Crash it
//
//  Created by Olivier Picard on 13/03/2018.
//  Copyright © 2018 Olivier Picard. All rights reserved.
//

import Foundation
import SpriteKit

class MovingItem: SKSpriteNode {
    static internal var base_moving_speed:Float = 1
    
    /** détermine les marges en dehors de l'écran à partir
     desquels on peut suprimer un élément
     
    - Note:
     Des items(ex : étoiles) peuvent être générer hors de l'écran proche du bord.
     Avant de dire qu'un item est hors de l'écran on vérfie qu'il soit très éloignés du bord de l'écran
     pour éviter la confusion avec les item récent et ceux en fin de vie d'où l'utiliter de définir
     cette variable pour avoir une marge fixe et connu
     */
    private let MARGIN_OUT_OF_SCREEN_TO_DELETE:CGFloat = 30
    
    /** determine the factor of the speed of this item (sprite) */
    internal var speed_factor: Float = 1
    /** determine the direction that this sprite are heading towards */
    internal var direction: CGVector
    
 
    
    init( texture: SKTexture?, color: UIColor, size: CGSize, speedFactor:Float = 2.5,
          direction: CGVector = CGVector(dx: 0, dy: -1)) {
        
        self.speed_factor = speedFactor
        self.direction = direction
        super.init(texture: texture, color: color, size: size)
    }
    
    
    convenience init(texture: SKTexture?, size: CGSize) {
        self.init(texture: texture, color: UIColor.white, size: size)
    }
    
    
    required init?(coder aDecoder: NSCoder) {
        self.speed_factor = 1
        self.direction = CGVector(dx: 0, dy: -1)
        super.init(coder: aDecoder)
    }
    
    /**
     Autorise l'item à se déplacer suivant la vitesse et la direction.
     - Important:
     il parait judicieux d'appeler cette fonction à chaque update de la scène
    */
    func update(_ currentTime: TimeInterval) {
        let realSpeed = CGFloat(MovingItem.base_moving_speed * self.speed_factor)
        let realDirection = CGVector(dx: direction.dx * realSpeed,
                                     dy: direction.dy * realSpeed)
        
        self.position = CGPoint(x: self.position.x + realDirection.dx,
                                y: self.position.y + realDirection.dy)
    }
    
    /**
     Détermine si l'item est hors de l'écran (n'est plus visible)
     - parameters:
        - screenSize : représente la taille de l'écran
     - returns:
     Retourne "true" si l'item est hors de l'écran sinon retourne "false"
     */
    func isOutOfScreen(screenSize: CGSize) -> Bool{
        // Anchor point 0.5 par défaut donc on divise la taille par 2
        if(self.position.y + self.size.height/2 < -MARGIN_OUT_OF_SCREEN_TO_DELETE
            || self.position.y - self.size.height/2 > screenSize.height + MARGIN_OUT_OF_SCREEN_TO_DELETE
            || self.position.x - self.size.width/2 > screenSize.width + MARGIN_OUT_OF_SCREEN_TO_DELETE
            || self.position.x + self.size.width/2 < -MARGIN_OUT_OF_SCREEN_TO_DELETE) {
            
            return true
        }
        
        return false
    }
    
    
    
    
}

File: Android/app/src/main/java/com/olivierpicard/crachit/MovingItem.java
package com.olivierpicard.crachit;

import android.support.annotation.Nullable;

import com.olivierpicard.crachit.Graphics.GSize;
import com.olivierpicard.crachit.Graphics.GSprite;
import com.olivierpicard.crachit.Graphics.GTools;
import com.olivierpicard.crachit.Graphics.GVector;
import com.olivierpicard.crachit.Graphics.IGUpdatable;

/**
 * Représente un item(sprite) qui se déplace sur l'écran
 * Created by olivierpicard on 03/04/2018.
 */

public class MovingItem extends GSprite implements IGUpdatable {
    static float base_moving_speed = 1.2f;
    private final int MARGIN_OUT_OF_SCREEN_TO_DELETE = 30;
    public float speed_factor = 1;
    public GVector direction;


    private void init() {
        this.speed_factor = 2.5f;
        this.direction = GVector.down();
    }

    public MovingItem(@Nullable Integer bitmapRessourceID, @Nullable Integer color, GSize size) {
        super(bitmapRessourceID, color, size);
        init();

    }


    public MovingItem(@Nullable Integer bitmapRessourceID, @Nullable Integer color, GSize size,
                      @Nullable Float speed_factor, GVector direction) {
        super(bitmapRessourceID, color, size);
        init();
        if(bitmapRessourceID != null) this.setBitmap(bitmapRessourceID);
        if(speed_factor != null) this.speed_factor = speed_factor;
        this.direction = direction;
    }



    public void update(long currentTime) {
        final float realSpeed = MovingItem.base_moving_speed * this.speed_factor;
        final GVector realDirection = new GVector(this.direction).multiply(realSpeed);
        setPosition(this.getPosition().add(realDirection));
    }


    public boolean isOutOfScreen() {

        // Anchor point 0.5 par défaut donc on divise la taille par 2
        if (getPosition().y + getSize().height / 2 < -MARGIN_OUT_OF_SCREEN_TO_DELETE
                || getPosition().y - getSize().height / 2 > this.getScene().getSize().height + MARGIN_OUT_OF_SCREEN_TO_DELETE
                || getPosition().x - getSize().width / 2 > this.getScene().getSize().width + MARGIN_OUT_OF_SCREEN_TO_DELETE
                || getPosition().x + getSize().width / 2 < -MARGIN_OUT_OF_SCREEN_TO_DELETE) {
            return true;
        }
        return false;
    }


    public boolean isOverlaps(ICollisionable item) {
        if(getPosition().x - getSize().width/2 < item.getPosition().x + item.getSize().width/2
                && getPosition().x + getSize().width/2 > item.getPosition().x - item.getSize().width/2
                && getPosition().y + getSize().height/2 > item.getPosition().y - item.getSize().height/2
                && getPosition().y - getSize().height/2 < item.getPosition().y + item.getSize().height/2) {
            return true;
        }
        return false;
    }
}
